### ISSUE TITLE
window.myFunction = function() triggers no-undef

### ISSUE THREAD

**Issue by ChildishGiant** [eslint/eslint#11968]
**What rule do you want to change?** [no-undef](https://eslint.org/docs/rules/no-undef)

**Does this change cause the rule to produce more or fewer warnings?** Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?** I would imagine most people would want it to be an option but I would be happy for this to be default behaviour.

**Please provide some example code that this change will affect:**

```js
window.myFunction = function () {
	console.log("Hello world!");
};

myFunction();
```

**What does the rule currently do for this code?** This throws `'myFunction' is not defined.`

**What will the rule do after it's changed?** This would pass the check.

**Are you willing to submit a pull request to implement this change?** I have not looked into the code base but could do in the future.

**Comment by platinumazure** [comment]
Hi @ChildishGiant, thanks for the issue.

In general, ESLint is not aware of the global object as the name of the global object varies from environment to environment. So I'm not sure this is a change we can safely make in the core `no-undef` rule.

In your case, you could declare `myFunction` as a global in your configuration or at the top of the script, which tells ESLint all it needs to know.

**Comment by g-plane** [comment]
As @platinumazure said, the code snippet will be as below:

```js
/* global myFunction */

window.myFunction = function () {
	console.log("Hello world!");
};

myFunction();
```

Here is [online demo](https://eslint.org/demo/#eyJ0ZXh0Ijoid2luZG93Lm15RnVuY3Rpb24gPSBmdW5jdGlvbiAoKSB7XHJcblx0Y29uc29sZS5sb2coXCJIZWxsbyB3b3JsZCFcIik7XHJcbn07XHJcblxyXG5teUZ1bmN0aW9uKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXVuZGVmIjoyfSwiZW52Ijp7ImJyb3dzZXIiOnRydWV9fX0=).

**Comment by ChildishGiant** [comment]
I did consider this as eslint isn't always used in a browser environment.  When playing around with the online demo I noticed that `/* global window */` doesn't fix the error. I'd like to be able to announce window as a global var to fix this for each instance with one stroke. I do see your side of the story though, so understand why this probably won't be changed, even as an option

**Comment by finico** [comment]
@ChildishGiant you can fix it in a config's `globals`
https://eslint.org/docs/user-guide/configuring#specifying-globals

**Comment by kaicataldo** [comment]
@ChildishGiant Specifying the `"browser"` env should get you the behavior you're after. Please see the [docs](https://eslint.org/docs/user-guide/configuring#specifying-environments) for more details.

**Comment by ChildishGiant** [comment]
![image](https://user-images.githubusercontent.com/13716824/61970272-5675d080-afd4-11e9-93cb-d857a332ec27.png)
I already have it set and the issue is still there

**Comment by kaicataldo** [comment]
Sorry, maybe I misunderstood your last comment. If you’re asking for the ability to not have to set arbitrary globals on the global, there’s no way for ESLint to know that it is defined or not.

I personally just always use the global object when defining and calling globals - it has the added side effect of making it very explicit!

**Comment by finico** [comment]
@ChildishGiant 

```json
{
    "env": {
        "browser": true,
        "es6": true
    },
    "parserOptions": {
        "ecmaVersion": 6
    },
    "globals": {
        "myFunction": "readonly"
    },
    "rules": {
        "no-undef": "error"
    }
}
```
